python/ray/workflow/http_event_provider.py

```diff
@@ -0,0 +1,240 @@
+        self.event_key_payload: Dict[str, Dict[Any, Any]] = {}
+        self.event_coordinator_ack = event_coordinator_ack
+        if self.event_coordinator_ack:
+            self.event_key_checkpointed: Dict[str, Dict[Any, Any]] = {}
+        else:
+            self.payload_record: Dict[Any] = {}
+
+    @app.post("/arrivals")
+    async def create_event_payload(self, req: Request):
+        req_json = await req.json()
+        workflow_id = req_json['workflow_id']
+        event_key = req_json['event_key']
+        event_payload = req_json['event_payload']
+        if workflow_id not in self.event_key_payload.keys():
+            return f"Event Dropped: not registered, please resend."
+        if event_key not in self.event_key_payload[workflow_id].keys():
+            return f"Event Dropped: not registered, please resend."
+        self.event_key_payload[workflow_id][event_key].set_result(event_payload)
+
+        if self.event_coordinator_ack:
+            # wait for event listener to report event has been checkpointed before acknowledging
+            if workflow_id not in self.event_key_checkpointed:
+                self.event_key_checkpointed[workflow_id] = {}
+            self.event_key_checkpointed[workflow_id][event_key] = asyncio.Future()
+            # await until confirmation of checkpointed is received.
+            confirmed = await self.event_key_checkpointed[workflow_id][event_key]
+            self.event_key_checkpointed[workflow_id].pop(event_key)
+            if confirmed:
+                return f"ACK_OK: event received and checkpointed successfully."
+            else:
+                # the ECA confirms the failure of checkpointing the event.
+                return f"ERROR: event received but failed in checkpointing."
+        else:
+            self.payload_record[event_payload] = asyncio.Future()
+            confirmed = await self.payload_record[event_payload]
+            self.payload_record.pop(event_payload)
+            if confirmed:
+                return f"ACK_OK: event received and checkpointed successfully."
+            else:
+                return f"ERROR: event received but failed in checkpointing."
+
+    async def get_event_payload(self, workflow_id, event_key):
+        if workflow_id not in self.event_key_payload.keys():
+            self.event_key_payload[workflow_id] = {}
+        if event_key in self.event_key_payload[workflow_id]:
+            raise WorkflowEventHandleError(workflow_id, f"The same {event_key} is used to get payload again.")
+        self.event_key_payload[workflow_id][event_key] = asyncio.Future()
+        return await self.event_key_payload[workflow_id][event_key]
+
+    async def report_checkpointed(self, event, confirmation):
+        self.payload_record[event].set_result(confirmation)
+
+    async def report_checkpointed_eca(self, workflow_id, event_key, confirmation):
+        self.event_key_checkpointed[workflow_id][event_key].set_result(confirmation)
```

Comment:
I'm feeling confused here. Do you know why do we have two methods here doing this? Also, without event coordinator, why there is no workflow_id?

From my understanding, this class should be shared by all workflows.